[HDR] The gain-map image should be decoded as HDR only if it is displayed on HDR display

shallawa opened the pull request:
#### 3343818fbd1f972e0adeebdd92da2328f0469e38
<pre>
[HDR] The gain-map image should be decoded as HDR only if it is displayed on HDR display
<a href="https://bugs.webkit.org/show_bug.cgi?id=294150">https://bugs.webkit.org/show_bug.cgi?id=294150</a>
<a href="https://rdar.apple.com/152744271">rdar://152744271</a>

Reviewed by Simon Fraser.

The gain map image decodes an SDR image and gain-map. The gain-map can only be
used on HDR display. Decoding the SDR image should be requested for SDR display
and for the CSS style &quot;dynamic-range-limit: standard;&quot; case.

RenderLayerBacking has to propagate its GraphicsLayer::drawsHDRContent() to the
painting code through the PaintBehavior. RenderImage will pass it to
BitmapImage::draw() through ImagePaintingOptions. It will be eventually passed
to ImageDecoder as ShouldDecodeToHDR though the DecodingOptions.

RenderLayerBacking should consider turning off the GraphicsLayer::drawsHDRContent()
which may happen when moving the WebView from an HDR display to an SDR display.

ImageFrame should hold two sources of images: one for SDR and the second for HDR.
ImageFrame::Source should hold: NativeImage, DecodingOptions and Headroom.
Managing the sources is controlled by an argument of type ShouldDecodeToHDR which
is now passed to most of ImageFrame methods.

* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::currentFrameHeadroom const): Deleted.
* Source/WebCore/loader/cache/CachedImage.h:
* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::draw):
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/BitmapImageSource.cpp:
(WebCore::BitmapImageSource::destroyNativeImageAtIndex):
(WebCore::BitmapImageSource::isCompatibleWithOptionsAtIndex const):
(WebCore::BitmapImageSource::decode):
(WebCore::BitmapImageSource::imageFrameDecodeAtIndexHasFinished):
(WebCore::BitmapImageSource::cacheNativeImageAtIndex):
(WebCore::BitmapImageSource::nativeImageAtIndexCacheIfNeeded):
(WebCore::BitmapImageSource::nativeImageAtIndexRequestIfNeeded):
(WebCore::BitmapImageSource::currentNativeImageForDrawing):
* Source/WebCore/platform/graphics/BitmapImageSource.h:
* Source/WebCore/platform/graphics/Image.h:
(WebCore::Image::hasHDRContent const):
(WebCore::Image::currentFrameHeadroom const): Deleted.
* Source/WebCore/platform/graphics/ImageDecoder.cpp:
(WebCore::ImageDecoder::fetchFrameMetaDataAtIndex const):
* Source/WebCore/platform/graphics/ImageDecoder.h:
(WebCore::ImageDecoder::frameOrientationAtIndex const):
(WebCore::ImageDecoder::frameHeadroomAtIndex const): Deleted.
* Source/WebCore/platform/graphics/ImageFrame.cpp:
(WebCore::ImageFrame::ImageFrame):
(WebCore::ImageFrame::operator=):
(WebCore::ImageFrame::clearSourceImage):
(WebCore::ImageFrame::clearImage):
(WebCore::ImageFrame::hasNativeImage const):
(WebCore::ImageFrame::hasFullSizeNativeImage const):
(WebCore::ImageFrame::hasDecodedNativeImageCompatibleWithOptions const):
* Source/WebCore/platform/graphics/ImageFrame.h:
(WebCore::ImageFrame::sizeInBytes const):
(WebCore::ImageFrame::setDensityCorrectedSize):
(WebCore::ImageFrame::densityCorrectedSize const):
(WebCore::ImageFrame::subsamplingLevel const):
(WebCore::ImageFrame::nativeImage const):
(WebCore::ImageFrame::decodingOptions const):
(WebCore::ImageFrame::headroom const):
(WebCore::ImageFrame::hasNativeImage const):
(WebCore::ImageFrame::Source::hasNativeImage const):
(WebCore::ImageFrame::Source::hasFullSizeNativeImage const):
(WebCore::ImageFrame::Source::hasDecodedNativeImageCompatibleWithOptions const):
(WebCore::ImageFrame::Source::clear):
(WebCore::ImageFrame::shouldDecodeToHDRIfExists const):
(WebCore::ImageFrame::source):
(WebCore::ImageFrame::source const):
(WebCore::ImageFrame::frameBytes const): Deleted.
(WebCore::ImageFrame::setHeadroom): Deleted.
(WebCore::ImageFrame::hasNativeImage): Deleted.
(WebCore::ImageFrame::hasFullSizeNativeImage): Deleted.
* Source/WebCore/platform/graphics/ImagePaintingOptions.h:
(WebCore::ImagePaintingOptions::drawsHDRContent const):
(WebCore::ImagePaintingOptions::setOption):
* Source/WebCore/platform/graphics/ImageSource.h:
(WebCore::ImageSource::shouldDecodeToHDR const): Deleted.
* Source/WebCore/platform/graphics/ImageTypes.h:
* Source/WebCore/platform/graphics/NativeImage.cpp:
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/graphics/NativeImageSource.h:
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::drawNativeImageInternal):
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
(WebCore::ImageDecoderCG::fetchFrameMetaDataAtIndex const):
(WebCore::headroomFromProperties): Deleted.
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
* Source/WebCore/rendering/PaintPhase.h:
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::paintIntoRect):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::PaintedContentRequest::PaintedContentRequest):
(WebCore::RenderLayer::paintLayerContents):
(WebCore::RenderLayer::paintForegroundForFragments):
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateDrawsContent):
(WebCore::RenderLayerBacking::paintContents):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/296430@main">https://commits.webkit.org/296430@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b9caaef66c66509278e8d99540bc305f087a2339

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108520 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28181 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18605 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113729 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58918 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110483 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28870 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36733 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82418 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111468 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22907 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97747 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62854 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22324 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15887 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58444 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92277 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15938 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116850 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35574 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26219 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91442 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35947 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94019 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91243 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23245 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36138 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13902 "Passed tests") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31323 "Hash b9caaef6 for PR 46450 does not build (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35476 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41011 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35186 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38537 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36866 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/DecodingOptions.h:
```diff
@@ -65,7 +65,7 @@ class DecodingOptions {
     bool isCompatibleWith(const DecodingOptions& other) const
     {
-        if (shouldDecodeToHDR() != other.shouldDecodeToHDR())
+        if (shouldDecodeToHDR() == ShouldDecodeToHDR::No && other.shouldDecodeToHDR() == ShouldDecodeToHDR::Yes)
```
The asymmetry here is surprising (and fragile).

smfr commented on Source/WebCore/platform/graphics/ImageFrame.h:
```diff
@@ -71,45 +71,85 @@ class ImageFrame {
     Seconds m_duration;
     bool m_hasAlpha { true };
+
+    std::array<Source, 2> m_sources;
```
You might just as well have
```
Source m_source;
Source m_hdrSource;
```

or
```
Source m_source;
std::optional<Source> m_hdrSource;
```

shallawa committed:
[HDR] The gain-map image should be decoded as HDR only if it is displayed on HDR display
https://bugs.webkit.org/show_bug.cgi?id=294150
rdar://152744271

Reviewed by Simon Fraser.

The gain map image decodes an SDR image and gain-map. The gain-map can only be
used on HDR display. Decoding the SDR image should be requested for SDR display
and for the CSS style "dynamic-range-limit: standard;" case.

RenderLayerBacking has to propagate its GraphicsLayer::drawsHDRContent() to the
painting code through the PaintBehavior. RenderImage will pass it to
BitmapImage::draw() through ImagePaintingOptions. It will be eventually passed
to ImageDecoder as ShouldDecodeToHDR though the DecodingOptions.

RenderLayerBacking should consider turning off the GraphicsLayer::drawsHDRContent()
which may happen when moving the WebView from an HDR display to an SDR display.

ImageFrame should hold two sources of images: one for SDR and the second for HDR.
ImageFrame::Source should hold: NativeImage, DecodingOptions and Headroom.
Managing the sources is controlled by an argument of type ShouldDecodeToHDR which
is now passed to most of ImageFrame methods.

* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::currentFrameHeadroom const): Deleted.
* Source/WebCore/loader/cache/CachedImage.h:
* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::draw):
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/BitmapImageSource.cpp:
(WebCore::BitmapImageSource::destroyNativeImageAtIndex):
(WebCore::BitmapImageSource::isCompatibleWithOptionsAtIndex const):
(WebCore::BitmapImageSource::decode):
(WebCore::BitmapImageSource::imageFrameDecodeAtIndexHasFinished):
(WebCore::BitmapImageSource::cacheNativeImageAtIndex):
(WebCore::BitmapImageSource::nativeImageAtIndexCacheIfNeeded):
(WebCore::BitmapImageSource::nativeImageAtIndexRequestIfNeeded):
(WebCore::BitmapImageSource::currentNativeImageForDrawing):
* Source/WebCore/platform/graphics/BitmapImageSource.h:
* Source/WebCore/platform/graphics/Image.h:
(WebCore::Image::hasHDRContent const):
(WebCore::Image::currentFrameHeadroom const): Deleted.
* Source/WebCore/platform/graphics/ImageDecoder.cpp:
(WebCore::ImageDecoder::fetchFrameMetaDataAtIndex const):
* Source/WebCore/platform/graphics/ImageDecoder.h:
(WebCore::ImageDecoder::frameOrientationAtIndex const):
(WebCore::ImageDecoder::frameHeadroomAtIndex const): Deleted.
* Source/WebCore/platform/graphics/ImageFrame.cpp:
(WebCore::ImageFrame::ImageFrame):
(WebCore::ImageFrame::operator=):
(WebCore::ImageFrame::clearSourceImage):
(WebCore::ImageFrame::clearImage):
(WebCore::ImageFrame::hasNativeImage const):
(WebCore::ImageFrame::hasFullSizeNativeImage const):
(WebCore::ImageFrame::hasDecodedNativeImageCompatibleWithOptions const):
* Source/WebCore/platform/graphics/ImageFrame.h:
(WebCore::ImageFrame::sizeInBytes const):
(WebCore::ImageFrame::setDensityCorrectedSize):
(WebCore::ImageFrame::densityCorrectedSize const):
(WebCore::ImageFrame::subsamplingLevel const):
(WebCore::ImageFrame::nativeImage const):
(WebCore::ImageFrame::decodingOptions const):
(WebCore::ImageFrame::headroom const):
(WebCore::ImageFrame::hasNativeImage const):
(WebCore::ImageFrame::Source::hasNativeImage const):
(WebCore::ImageFrame::Source::hasFullSizeNativeImage const):
(WebCore::ImageFrame::Source::hasDecodedNativeImageCompatibleWithOptions const):
(WebCore::ImageFrame::Source::clear):
(WebCore::ImageFrame::shouldDecodeToHDRIfExists const):
(WebCore::ImageFrame::source):
(WebCore::ImageFrame::source const):
(WebCore::ImageFrame::frameBytes const): Deleted.
(WebCore::ImageFrame::setHeadroom): Deleted.
(WebCore::ImageFrame::hasNativeImage): Deleted.
(WebCore::ImageFrame::hasFullSizeNativeImage): Deleted.
* Source/WebCore/platform/graphics/ImagePaintingOptions.h:
(WebCore::ImagePaintingOptions::drawsHDRContent const):
(WebCore::ImagePaintingOptions::setOption):
* Source/WebCore/platform/graphics/ImageSource.h:
(WebCore::ImageSource::shouldDecodeToHDR const): Deleted.
* Source/WebCore/platform/graphics/ImageTypes.h:
* Source/WebCore/platform/graphics/NativeImage.cpp:
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/graphics/NativeImageSource.h:
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::drawNativeImageInternal):
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
(WebCore::ImageDecoderCG::fetchFrameMetaDataAtIndex const):
(WebCore::headroomFromProperties): Deleted.
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
* Source/WebCore/rendering/PaintPhase.h:
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::paintIntoRect):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::PaintedContentRequest::PaintedContentRequest):
(WebCore::RenderLayer::paintLayerContents):
(WebCore::RenderLayer::paintForegroundForFragments):
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateDrawsContent):
(WebCore::RenderLayerBacking::paintContents):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/296430@main
